src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -466,20 +407,19 @@ public Task<List<StreamSubscription>> GetAllSubscriptions(StreamId streamId, ISt
             {
                 RemoveProducer(producer);
             }
-        }
-
-        /// <summary>
-        /// Try clear state will only clear the state if there are no producers or consumers.
-        /// </summary>
-        /// <returns></returns>
-        private async Task<bool> TryClearState()
-        {
-            if (State.Producers.Count == 0 && State.Consumers.Count == 0) // + we already know that numProducers == 0 from previous if-clause
+            catch (OrleansMessageRejectionException)
             {
-                await ClearStateAsync(); //State contains no producers or consumers, remove it from storage
-                return true;
+                var grainRef = producer.Producer as GrainReference;
+                // if producer is a system target on and unavailable silo, remove it.
+                if (grainRef == null || grainRef.GrainId.IsSystemTarget && siloStatusOracle.GetApproximateSiloStatus(grainRef.SystemTargetSilo).IsUnavailable())
+                {
+                    RemoveProducer(producer);
+                }
+                else // otherwise, throw
+                {
+                    throw;
+                }
             }
-            return false;
```

Comment:
There's a case where Silo A messages Silo B (who does not know about Silo A yet), Silo A dies. Since Silo B never knew about Silo A, Service Fabric membership will never inform it of A's demise. SF membership works based upon progressive consistent snapshots of cluster state and does not include a history of all silos which ever existed.

Anyway: I believe this is fine because eventually the call will fail in `PrepareMessageForSend` under the `lastConnectionFailure` clause as long as messages happen in quick enough succession.

This is worth testing, perhaps via a test which uses a mocked membership oracle.